modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -74,27 +86,45 @@ public TxManagerImpl(ReplicaService replicaService, LockManager lockManager, Hyb
         this.clock = clock;
     }
 
-    /** {@inheritDoc} */
     @Override
     public InternalTransaction begin() {
         return begin(false);
     }
 
-    /** {@inheritDoc} */
     @Override
     public InternalTransaction begin(boolean readOnly) {
         UUID txId = Timestamp.nextVersion().toUuid();
 
-        return readOnly ? new ReadOnlyTransactionImpl(this, txId, clock.now()) : new ReadWriteTransactionImpl(this, txId);
+        if (!readOnly) {
+            return new ReadWriteTransactionImpl(this, txId);
+        }
+
+        HybridTimestamp readTimestamp = clock.now();
+
+        readOnlyTxFutureByReadTs.compute(readTimestamp, (timestamp, readOnlyTxFuture) -> {
+            assert readOnlyTxFuture == null : "previous transaction has not completed yet: " + readTimestamp;
+
+            HybridTimestamp lowerBound = lowerBoundTsToStartNewReadOnlyTx.get();
+
+            if (lowerBound != null && readTimestamp.compareTo(lowerBound) <= 0) {
+                throw new IgniteInternalException(
+                        TX_READ_ONLY_CREATING_ERR,
```

Comment:
Do we need such a general error cause ('some error while creating an RO transaction'? Won't it make sense to reserve a specific error code for this specific error: 'RO read tx is too old'?

Reply:
Why not.